ucsmsdk/utils/ucstechsupport.py

```diff
@@ -19,11 +19,258 @@
 import time
 import datetime
 import logging
+from ..deprecated import deprecated
 from ..ucsexception import UcsValidationException, UcsWarning
+from ..mometa.sysdebug.SysdebugTechSupportCmdOpt import \
+    SysdebugTechSupportCmdOptConsts as TechSupOptsConsts
 
 log = logging.getLogger('ucs')
 
 
+def _create_download_dir(file_dir):
+    if not os.path.exists(file_dir):
+        os.makedirs(file_dir)
+
+
+def _validate_download_args(file_dir, file_name):
+    if file_dir is None:
+        raise UcsValidationException('file_dir is a required parameter')
+
+    if file_name is None:
+        raise UcsValidationException('file_name is a required parameter')
+
+    if not file_name.endswith('.tar'):
+        raise UcsValidationException('file_name should end with .tar')
+
+
+def _get_creation_ts():
+    dt1 = datetime.datetime(1970, 1, 1, 12, 0, 0, 0)
+    dt2 = datetime.datetime.utcnow()
+    return int((dt2 - dt1).total_seconds())
+
+
+def _bootstrap_tech_support_mo(creation_ts):
+    from ..mometa.top.TopSystem import TopSystem
+    from ..mometa.sysdebug.SysdebugTechSupport import SysdebugTechSupport, \
+        SysdebugTechSupportConsts
+    from ..mometa.sysdebug.SysdebugTechSupFileRepository import \
+        SysdebugTechSupFileRepository
+
+    top_system = TopSystem()
+    ts_file_repo = SysdebugTechSupFileRepository(parent_mo_or_dn=top_system)
+    ts_mo = SysdebugTechSupport(
+        parent_mo_or_dn=ts_file_repo,
+        creation_ts=str(creation_ts),
+        admin_state=SysdebugTechSupportConsts.ADMIN_STATE_START)
+    return ts_mo
+
+
+def _is_valid_arg(param, kwargs):
+    return param in kwargs and kwargs[param] is not None
+
+
+def _set_ts_options_ucsm(ts_cmd_opt, kwargs):
+    ts_cmd_opt.major_opt_type = TechSupOptsConsts.MAJOR_OPT_TYPE_UCSM
+
+
+def _set_ts_options_ucsm_mgmt(ts_cmd_opt, kwargs):
+    ts_cmd_opt.major_opt_type = TechSupOptsConsts.MAJOR_OPT_TYPE_UCSM_MGMT
+
+
+def _validate_chassis_options(kwargs):
+    if not _is_valid_arg("chassis_id", kwargs):
+        raise UcsValidationException('chassis_id should be specified')
+    if "cimc_id" not in kwargs and "iom_id" not in kwargs:
+        raise UcsValidationException('cimc_id/iom_id should be specified')
+
+
+def _set_ts_options_chassis(ts_cmd_opt, kwargs):
+    _validate_chassis_options(kwargs)
+    ts_cmd_opt.chassis_id = str(kwargs["chassis_id"])
+    ts_cmd_opt.major_opt_type = TechSupOptsConsts.MAJOR_OPT_TYPE_CHASSIS
+    if _is_valid_arg("cimc_id", kwargs):
+        ts_cmd_opt.chassis_cimc_id = str(kwargs["cimc_id"])
+
+        if _is_valid_arg("adapter_id", kwargs):
```

Comment:
adapter_id is valid only if cimc_id is provided for all cimc_id adapter id is always ALL. So it can be:
if ts_cmd_opt.chassis_cimc_id != "all" and _is_valid_arg("adapter_id", kwargs):